### PR TITLE
Add support for running test classes in Metals

### PR DESCRIPTION
**Pull request by lwronski**
https://scala-cli.virtuslab.org/docs/cookbooks/vscode

### CODE REVIEW

File: modules/build/src/main/scala/scala/build/bsp/BspImpl.scala [hunk 1]
```diff
@@ -64,62 +68,106 @@ final class BspImpl(
...
-      preBuildData.generatedSources,
-      preBuildData.buildOptions,
+      preBuildDataMain.generatedSources,
+      preBuildDataMain.buildOptions,
       Scope.Main,
       logger,
       actualLocalClient,
       bloopServer
-    )
+    ).swap.map(e => (e, Scope.Main)).swap
```
Comment (romanowski) [inline modules/build/src/main/scala/scala/build/bsp/BspImpl.scala:159]:
```suggestion
    ).left.map(e => (e, Scope.Main))
```

File: modules/build/src/main/scala/scala/build/bsp/BspImpl.scala [hunk 2]
```diff
@@ -64,62 +68,106 @@ final class BspImpl(
...
+      inputs,
+      preBuildDataTest.sources,
+      inputs.generatedSrcRoot(Scope.Test),
+      preBuildDataTest.generatedSources,
+      preBuildDataTest.buildOptions,
+      Scope.Test,
+      logger,
+      actualLocalClient,
+      bloopServer
+    ).swap.map(e => (e, Scope.Test)).swap
```
Comment (romanowski) [inline modules/build/src/main/scala/scala/build/bsp/BspImpl.scala:170]:
```suggestion
    ).left.map(e => (e, Scope.Test))
```

File: modules/build/src/main/scala/scala/build/bsp/BspImpl.scala [hunk 3]
```diff
@@ -64,62 +68,106 @@ final class BspImpl(
...
+        generatedSourcesMain,
+        options0Main,
         Scope.Main,
         logger,
         localClient
-      )
+      ) match {
+        case Right(v) => Right(v)
+        case Left(e)  => Left((e, Scope.Main))
+      }
```
Comment (romanowski) [inline modules/build/src/main/scala/scala/build/bsp/BspImpl.scala:98]:
```suggestion
      ).left.map(e =>  Left((e, Scope.Main)))
```

File: modules/build/src/main/scala/scala/build/bsp/BspImpl.scala [hunk 4]
```diff
@@ -64,62 +68,106 @@ final class BspImpl(
...
+        sourcesTest,
+        generatedSourcesTest,
+        options0Test,
+        Scope.Test,
+        logger,
+        localClient
+      ) match {
+        case Right(v) => Right(v)
+        case Left(e)  => Left((e, Scope.Main))
+      }
```
Comment (romanowski) [inline modules/build/src/main/scala/scala/build/bsp/BspImpl.scala:113]:
```suggestion
      ).left.map(e =>  Left((e, Scope.Main)))
```

File: modules/integration/src/test/scala/scala/cli/integration/BspTestDefinitions.scala [hunk 5]
```diff
@@ -748,14 +758,100 @@ abstract class BspTestDefinitions(val scalaVersionOpt: Option[String])
...
+
+        val newFileContent =
+          """object Messages {
+            |  def msg = "Hello2"
+            |}
+            |""".stripMargin
+        os.write.over(root / "Messages.scala", newFileContent)
+
+        {
+          val resp = await(remoteServer.buildTargetCompile(new b.CompileParams(targets)).asScala)
```
Comment (romanowski) [inline modules/integration/src/test/scala/scala/cli/integration/BspTestDefinitions.scala:849]: I think we should also check negative path, by e.g. removing `def msg` and expecting compilation error

**Review by romanowski — APPROVED** [review]
I miss the tests to ensure that Metals are able to detect tests / main classes.

In existing tests, we do miss a negative scenario.

Few nits. 

For now we can merge it, with improved testing in a follow up

**Comment by ches** [comment]
What does it mean to "run test classes in Metals"? I didn't understand the feature from the v0.0.8 release notes, nor the PR description, and no doc updates in the PR. Not being very familiar with the code base, even reading the PR diff isn't much help…

**Comment by lwronski** [comment]
Hi @ches,
It means, that in VSCode you can run tests by pressing `test/debug test` button in test files. [Here](https://scala-cli.virtuslab.org/docs/cookbooks/vscode) is described simple use cases.

**Pull request merged**